runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractQueuedStreamView.java

```diff
@@ -626,52 +629,153 @@ public synchronized long find(long globalAddress, SearchDirection direction) {
         return result == null ? Address.NOT_FOUND : result;
     }
 
-    protected BackpointerOp resolveCheckpoint(final QueuedStreamContext context, ILogData data,
-                                              long maxGlobal) {
+    // Contains a list of all valid checkpoints for the given global max.
+    // We need to select the checkpoint with the highest checkpointed space
+    // (which does not necessarily correspond to its order of occurrence in the log).
+    private Map<UUID, StreamCheckpoint> checkpointCandidates = new HashMap<>();
+    private Checkpoint currentCheckpoint = new Checkpoint();
+
+    /**
+     * This class represents a checkpoint in its simplest form.
+     * It is represented by the checkpoint identifier and the address up to which
+     * this checkpoint has coalesced the state for the current stream.
+     *
+     * It is used to keep the highest checkpoint for the current stream.
+     **/
+    @Data
+    class Checkpoint {
+        private UUID checkpointId = null;
+        private long checkpointedStartAddress = Address.NEVER_READ;
+
+        public boolean validateHighest(UUID id, long startAddress) {
+            if (Address.isAddress(startAddress) && startAddress > checkpointedStartAddress) {
+                 checkpointedStartAddress = startAddress;
+                 checkpointId = id;
+                 log.trace("validateHighest[{}]: valid checkpoint {} with start address {}", this, id, startAddress);
+                 return true;
+            }
+            return false;
+        }
+
+        public void reset() {
+            checkpointId = null;
+            checkpointedStartAddress = Address.NEVER_READ;
+        }
+    }
+
+    /**
+     * Resolve all potential checkpoints for the given max global.
```

Comment:
Do we need to resolve all the checkpoints with end address lesser than given max global? Can we return as soon as we find start address of a single checkpoint while traversing from the end?

Reply:
No, we actually want to traverse all valid checkpoints, because an older checkpoint might appear later in the stream. Therefore, even this is the first we find it is not the highest one.